static/src/javascripts-legacy/projects/common/utils/check-mediator.js

```diff
@@ -0,0 +1,146 @@
+define([
+    'Promise'
+    ], function (
+    Promise
+) {
+
+    /**
+     * registeredChecks will store references to instances of DefferedCheck
+    **/
+    var registeredChecks = {};
+
+    /**
+     * referenced as passCondition for dependentChecks
+     * SOMECHECKSPASSED: Every dependentCheck has returned true
+     * EVERYCHECKPASSED: At least one dependentCheck has returned true
+    **/
+    var SOMECHECKSPASSED = Array.prototype.some;
+    var EVERYCHECKPASSED = Array.prototype.every;
+    /**
+     * checkList is an array of object literals.
+     * Each object in this array will be converted to a DefferedCheck and added to registeredChecks
+     * Each object contains 3 fields: id (string), canRun (boolean), dependentChecks (object)
+     * dependentChecks should contain 2 fields: passCondition (SOMECHECKSPASSED/EVERYCHECKPASSED), list (nested array of checks)
+     * If object has dependentChecks then the DefferedCheck will resolve when these dependentChecks have all resolved
+     *
+    **/
+
+    var checkList = [{
+            id: 'isOutbrainNonCompliant',
+            dependentChecks: {
+                passCondition: SOMECHECKSPASSED,
+                list:[{
+                    id: 'isUserInAClashingAbTest'
+                }, {
+                    id: 'canEmailBeInserted',
+                    dependentChecks: {
+                        passCondition: EVERYCHECKPASSED,
+                        list: [{
+                            id: 'emailCanRun'
+                        }, {
+                            id: 'listCanRun'
+                        }, {
+                            id: 'emailInArticleOutbrainEnabled'
+                        }]
+                    }
+                }]
+            }
+        }];
+
+    function DeferredCheck(dependentCheckPromises, dependentChecksPassCondition) {
+        this.complete = new Promise(function(resolve, reject) {
+            this.resolve = resolve;
+            this.reject = reject;
+        }.bind(this));
+
+        if (dependentCheckPromises) {
+            Promise.all(dependentCheckPromises).then(function(results) {
+
+                var hasPassed = function(result) {
+                    return result;
+                };
+
+                if (dependentChecksPassCondition.call(results, hasPassed)) {
+                    this.resolve(true);
+                } else {
+                    this.resolve(false);
+                }
+            }.bind(this));
+        }
+    }
+
+    function registerDefferedCheck(check) {
+        var dependentCheckPromises = [];
+
+        if (check.dependentChecks) {
+            check.dependentChecks.list.forEach(registerDependentCheck.bind(null, dependentCheckPromises));
```

Comment:
Can you use `'lodash/collections/map'` here instead of side-effecting the `dependentCheckPromises` array? If I am reading right it could then become:

```js
function registerDefferedCheck(check) {
        
        if (check.dependentChecks) {
          return new DeferredCheck(map(check.dependentChecks.list, registerDependentCheck), check.dependentChecks.passCondition);
        }

        return new DeferredCheck();
    }


    function registerDependentCheck(dependentCheck) {
        return registerCheck(dependentCheck).complete;
    }
```